railties/lib/rails/generators/generated_attribute.rb

```diff
@@ -88,26 +91,25 @@ def reference?(type)
         private
           # parse possible attribute options like :limit for string/text/binary/integer, :precision/:scale for decimals or :polymorphic for references/belongs_to
           # when declaring options curly brackets should be used
-          def parse_type_and_options(type_with_possible_options)
-            return type_with_possible_options, {} unless type_with_possible_options&.match?(TYPE_WITH_OPTIONS_RE)
-
-            case type_with_possible_options
-            when /(string|text|binary|integer)\{(\d+)\}/
-              return $1, limit: $2.to_i
-            when /decimal\{(\d+)[,.-](\d+)\}/
-              return :decimal, precision: $1.to_i, scale: $2.to_i
+          def parse_type_and_options(type, options)
+            return type, {} if options.blank?
+
+            if type.presence_in(%w(string text binary integer)) && options.match?(/^\d+$/)
+              return type.to_sym, limit: options.to_i
+            elsif type.presence_in(%w(decimal numeric)) && options.match?(/^(\d+)[,.-](\d+)$/)
+              precision, scale = options.split(/[,.-]/)
+              return :decimal, precision: precision.to_i, scale: scale.to_i
             end
 
-            type, provided_options = type_with_possible_options.match(TYPE_WITH_OPTIONS_RE).captures
-            attr_options = parse_attr_options(provided_options)
+            attr_options = parse_attr_options(options)
 
             return type, Hash[attr_options].deep_symbolize_keys
```

Comment:
Rails generally prefers case statements or nested if/elses instead of early returns, so I'd probably rewrite to:

```ruby
def parse_type_and_options(type, options)
  case
  when options.blank?
    [ type, {} ]
  when %w[string text binary integer].include?(type) && options.match?(/^\d+$/)
    [ type.to_sym, limit: options.to_i ]
  when %w[decimal numeric].include?(type) && options.match?(/^(\d+)[,.-](\d+)$/)
    precision, scale = options.split(/[,.-]/)
    [ :decimal, precision: precision.to_i, scale: scale.to_i ]
  else
    [ type, parse_attr_options(options).to_h.deep_symbolize_keys ]
  end
end
```

Reply:
Also we can just use `include?` here instead of `presence_in` since we don't care about having the type returned.

Reply:
f0a91e2c33